test/e2e/custom_wait_rules_test.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestCustomWaitRules(t *testing.T) {
+	env := BuildEnv(t)
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, env.KappBinaryPath, logger}
+	//kubectl := Kubectl{t, env.Namespace, logger}
+
+	config := `
+apiVersion: kapp.k14s.io/v1alpha1
+kind: Config
+
+waitRules:
+  - ytt:
+      waitRuleContractV1:
+        rules.star: |
+          load("@ytt:data", "data")
+
+          state = data.values.status.currentState
+
+          def check_status():
+              if state == "Failed":
+                return {"Done":True, "Successful": False, "Message": ""}
+              elif state == "Running":
+                return {"Done":True, "Successful": True, "Message": ""}
+              else:
+                return {"Done":False, "Successful": False, "Message": "Not in Failed or Running state"}
+              end
+          end
```

Comment:
@jtigger any opinions on this interface presented via kapp (e.g. should return value be something else than hash, etc.)?

Reply:
_(I'm mindful that this feedback is "rather late to the party" ... my hope is that it's helpful...)_

## Hook in the form of a `ytt` Library for Future Value?

I'm interested in the possibilities kept open by the content of `waitRuleContractV1` being a `ytt` Library (i.e. collection of files that are executed together).

_However_, doing so brings some ceremony that the user can't value: the user is naming _a lot_ of objects (a file, a function, unpacking the `state` variable) which they have no other need to do other than satisfy this interface.

Have we considered a simple interface where the value of the key is the body of a function?


## Return value of `check_status()` through a function?

Given that the desired output is _not_ arbitrary, but very specific in shape (i.e. is equally part of the contract), using an open-ended data structure like a hash invites the possibility of errors without much upside.

Would it be worth it to provide a function to format the result?

```python
load("@kapp:status", "status")
state = status.currentState
def check_status():
  if state == "Failed":
    return status.Result(done=True, successful=False, Message="")
  elif state == "Running":
    return status.Result(done=True, successful=True, Message="")
  else:
    return status.Result(done=False, successful=False, Message="Not in Failed or Running state")
  end
end
```

## Separate code declaration from reference?

If I'm reading this right, the ytt Library declaration _is_ also the reference.
So, if I want to declare multiple custom wait rules, I need to repeat that library structure.

I am imagining that the user might want to provide multiple custom wait rules. Would it help to extract the library into a separate value?

```yaml
waitRuleYttLibrary:
  rules.star: |
    ...
    def crontab_status():
    ...
waitRules:
- resourceMatchers:
  - apiVersionKindMatcher: {apiVersion: corp.com/v1, kind: DatabaseInstance}
  conditionMatchers:
  - type: Failed
    status: "True"
    failure: true
  - type: Deployed
    status: "True"
    success: true
  supportsObservedGeneration: true
- resourceMatchers:
   - apiVersionKindMatcher: {apiVersion: stable.example.com/v1, kind: CronTab}
   customStateMatcher: "crontab_status"
```

Reply:
Hi @jtigger

Thanks for this much detailed feedback  💯 . _(You are definitely not late to the party yet 😄 )_

> However, doing so brings some ceremony that the user can't value: the user is naming a lot of objects (a file, a function, unpacking the state variable) which they have no other need to do other than satisfy this interface.

Totally Agree with this part. 

> `load("@kapp:status", "status")`

Not clear about this part how we can import `status` from `@kapp:status`  and what is `@kapp:status` here

Reply:
> - resourceMatchers:
     - apiVersionKindMatcher: {apiVersion: stable.example.com/v1, kind: CronTab}
     customStateMatcher: "crontab_status"
   
   Here `crontab_status` is in quotes by this way `crontab_status` function will get called and return the result ?

Reply:
> Not clear about this how we can import `status` from `@kapp:status` and what is `@kapp:status` here ?

Yeah, apologies, I'm working off of some ideas that haven't yet been broadly considered.

Today, `ytt` comes with what you could think of as its "standard library". We call it `@ytt`.

A Library (in this sense) is a collection of "modules" (that's a Starlark concept: https://github.com/google/starlark-go/blob/master/doc/spec.md#module-execution). Examples are: `@ytt:data` is the "data" module, `@ytt:template` is the "template" module.

We don't have the full facility to do this in `ytt`, yet, but... I was imaging that we might collect one or more modules that are specifically in support of Starlark as it appears in `kapp`.
So as an aspirational sketch, I'm saying, "what if ytt allowed you to register your own library as well... and further, what if `kapp` _did_ register its own library... and we called it `@kapp`?"

`@kapp:status` could be the first of such modules in the `@kapp` ytt library.
This module could have utilities for presenting and processing resource status.

Now, under the covers, it may well be that the means by which we plumb the status of resource(s) _is_ as data values. And that makes a ton of sense to do so.
But I'm suggesting that in this particular context, that might be an implementation detail that the user neither cares about nor needs to understand. And would be the implementation _within_ this `@kapp:status` module.

Coincidentally, that module also can provide functions like `status.Result()` that helps catch/prevent syntactic errors and overall provide a better experience.

Reply:
I previously suggested:
> Have we considered a simple interface where the value of the key is the body of a function?

But on second thought, I'm seeing that limiting the "viewport" of this field to being only the inside of a function would cut-off opportunities to gracefully handle more complex cases.
Namely, ensuring the user continues to have the ability to load `@ytt:` modules will be valuable: there's facilities for encoding/decoding, parsing, and manipulating values.

Also, I could imagine that as adoption of a feature like this increases, it's not unreasonable to expect a need to easily share/reuse such rules (e.g. `vendir sync'ed` and referenced via a path rather than inline) would emerge.
So, making the unit of `ytt` code inclusion being a proper ytt library is making more sense, to me.

Reply:
BTW, I'm realizing that I've been using a term that's rather overloaded: "library".

In `ytt` there are three meanings of this word (sigh):
- the set of files/directories that you `-f` into `ytt` is known as a "Library".
  - when I'm musing about what shape should the user's input be: a function body? a module? a library? — I'm using _this_ definition of library, here.
  - note that we tend to only talk about so-called "private libraries", like here: [What is a library?](https://carvel.dev/ytt/docs/v0.40.0/lang-ref-ytt-library/#what-is-a-library); but note the mention of the "root library" in that description.
- a collection of loadable modules is _also_ known as a library: `@ytt:` is the primary (and currently, only) example here (it's this: [Built-in ytt library](https://carvel.dev/ytt/docs/v0.40.0/lang-ref-ytt/#docs))
  - when talking about this thing `@kapp:` _this_ is the definition of library I'm using there.
- when a file has the `.lib.yaml` filename extension, we colloquially refer to that as a "library file" (mostly because of that `.lib.` extension). It's more accurate to call it a "loadable module". (as described in [Load Statements](https://carvel.dev/ytt/docs/v0.40.0/lang-ref-load/))
  - this definition hasn't come up in this convo; just adding it for completeness.
 
Hopefully, this helps! 😬